### PR TITLE
implement configurable timeout for OBD tests 

### PR DESCRIPTION
**Pull request by wlan0**
to make sure it doesn't time out in case of large Minio clusters

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#e30d7ad7-19ce-4c38-b024-2489efd6c8dd) |
<a name="e30d7ad7-19ce-4c38-b024-2489efd6c8dd"></a>
### 9324-6e911fe/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:31976
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 79edb277fcd2:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 9 minutes and 36 seconds
(4/15) Running aws-sdk-ruby tests ... done in 17 seconds
(5/15) Running awscli tests ... FAILED in 1 minutes and 10 seconds
{
  "name": "awscli",
  "duration": 2120,
  "function": "aws --endpoint-url http://minio-dev3.minio.io:31976 s3api create-multipart-upload --bucket \nAn error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. --key \nAn error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it.-object\n",
  "status": "FAIL",
  "error": ""
}

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
One of the reasons why I asked if this should be configurable, can we do this now ?

**Review by harshavardhana — CHANGES_REQUESTED** [review]
We also need to now change the obd handler to make sure to keep writing data on the wire so the HTTP client doesn't timeout. 

For that, you should start using the TraceHandler style keepAlive ticker to maintain the connection

```
        keepAliveTicker := time.NewTicker(500 * time.Millisecond)
        defer keepAliveTicker.Stop()

        enc := json.NewEncoder(w)
        for {
                select {
                case entry := <-traceCh:
                        if err := enc.Encode(entry); err != nil {
                                return
                        }
                        w.(http.Flusher).Flush()
                case <-keepAliveTicker.C:
                        if _, err := w.Write([]byte(" ")); err != nil {
                                return
                        }
                        w.(http.Flusher).Flush()
                case <-GlobalServiceDoneCh:
                        return
                }
        }
```

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#b3804621-80e8-4a45-9b4f-0ce148b5a092) |
<a name="b3804621-80e8-4a45-9b4f-0ce148b5a092"></a>
### 9324-376c80d/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30612
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 93641d641fec:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 9 minutes and 45 seconds
(4/15) Running aws-sdk-ruby tests ... done in 18 seconds
(5/15) Running awscli tests ... done in 2 minutes and 54 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 6 minutes and 5 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 30 seconds
(9/15) Running minio-go tests ... done in 19 minutes and 21 seconds
(10/15) Running minio-java tests ... done in 27 minutes and 8 seconds
(11/15) Running minio-js tests ... FAILED in 3 minutes and 50 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-d2d8d31b-20f3-4948-a892-06dada941b1b, recursive:true",
  "duration": 210,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by wlan0** [comment]
> One of the reasons why I asked if this should be configurable, can we do this now ?

yeah this would be useful now

**Comment by wlan0** [comment]
@harshavardhana 

- I've added the deadline 
- Removed LockOSThread

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#048058d6-addb-441d-b1dc-cb08cbd1cb61) |
<a name="048058d6-addb-441d-b1dc-cb08cbd1cb61"></a>
### 9324-46129c1/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31132
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9f7133ecca67:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 15 seconds
(4/15) Running aws-sdk-ruby tests ... done in 15 seconds
(5/15) Running awscli tests ... done in 3 minutes and 10 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 6 minutes and 38 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 34 seconds
(9/15) Running minio-go tests ... done in 21 minutes and 38 seconds
(10/15) Running minio-java tests ... done in 26 minutes and 38 seconds
(11/15) Running minio-js tests ... FAILED in 3 minutes and 47 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-94986fb5-643c-4e29-9ba8-f7519fde81cd, recursive:true",
  "duration": 210,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#bee3a8f8-646d-4095-b501-c317401990a4) |
<a name="bee3a8f8-646d-4095-b501-c317401990a4"></a>
### 9324-9e28cf5/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32458
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fc1e4989232f:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 16 seconds
(4/15) Running aws-sdk-ruby tests ... done in 15 seconds
(5/15) Running awscli tests ... done in 3 minutes and 1 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 6 minutes and 3 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 30 seconds
(9/15) Running minio-go tests ... done in 20 minutes and 0 seconds
(10/15) Running minio-java tests ... done in 26 minutes and 4 seconds
(11/15) Running minio-js tests ... FAILED in 3 minutes and 46 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-ecb9fd35-68a0-4454-91af-6ed292ae1a71, recursive:true",
  "duration": 213,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/admin-handlers.go [hunk 1]
```diff
@@ -1199,15 +1200,51 @@ func (a adminAPIHandlers) OBDInfoHandler(w http.ResponseWriter, r *http.Request)
...
 
+	deadlineSeconds := time.Duration(3600)
+	deadline := r.URL.Query().Get("deadlineseconds")
+	if deadline != "" {
+		deadlineSecondsInt, err := strconv.Atoi(deadline)
+		if err != nil {
+			errResp(err)
+			return
+		}
+		deadlineSeconds = time.Duration(deadlineSecondsInt)
```
Comment (poornas) [inline cmd/admin-handlers.go:1211]: perhaps [time.ParseDuration](https://golang.org/pkg/time/#ParseDuration) would be more idiomatic

File: cmd/admin-handlers.go [hunk 2]
```diff
@@ -1199,15 +1200,51 @@ func (a adminAPIHandlers) OBDInfoHandler(w http.ResponseWriter, r *http.Request)
...
+				doPartialWrite()
+			case <-pulse:
+				doPartialWrite()
+			case <-obdDone:
+				break
+			case <-deadlinedCtx.Done():
+				break
+			}
+		}
+		doPartialWrite()
```
Comment (poornas) [inline cmd/admin-handlers.go]: travis build failure here as code is unreachable.

File: pkg/madmin/obd.go [hunk 3]
```diff
@@ -221,11 +222,13 @@ var OBDDataTypesList = []OBDDataType{
 
 // ServerOBDInfo - Connect to a minio server and call OBD Info Management API
 // to fetch server's information represented by OBDInfo structure
-func (adm *AdminClient) ServerOBDInfo(ctx context.Context, obdDataTypes []OBDDataType) <-chan OBDInfo {
+func (adm *AdminClient) ServerOBDInfo(ctx context.Context, obdDataTypes []OBDDataType, deadlineSeconds time.Duration) <-chan OBDInfo {
 	respChan := make(chan OBDInfo)
 	go func() {
 		v := url.Values{}
 
+		v.Set("deadlineseconds", fmt.Sprintf("%v", deadlineSeconds.Truncate(1*time.Second).Seconds()))
```
Comment (poornas) [inline pkg/madmin/obd.go]:
```suggestion
		v.Set("deadlineseconds", deadlineSeconds.Truncate(1*time.Second).String())
```

**Review by poornas — CHANGES_REQUESTED** [review]

File: pkg/madmin/obd.go (hunk 3 above)
Comment (wlan0) [reply pkg/madmin/obd.go]: the `.String()` for `time.Duration` uses a human readable format, like `30m20s`.

File: cmd/admin-handlers.go (hunk 2 above)
Comment (wlan0) [reply cmd/admin-handlers.go]: this is fixed now

File: pkg/madmin/obd.go (hunk 3 above)
Comment (wlan0) [reply pkg/madmin/obd.go]: and I just realized, I'm actually printing the output of `.Seconds()`, which is float64

Comment (harshavardhana) [reply pkg/madmin/obd.go]: yeah, you can directly use. 
```
v.Set("deadlineseconds", deadlineSeconds.Truncate(1*time.Second).String())
```

File: cmd/admin-handlers.go [hunk 4]
```diff
@@ -1199,15 +1200,52 @@ func (a adminAPIHandlers) OBDInfoHandler(w http.ResponseWriter, r *http.Request)
 		finish()
 	}
 
+	deadlineSeconds := time.Duration(3600)
+	deadline := r.URL.Query().Get("deadlineseconds")
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: if we did stringification then we can simply use `time.ParseDuration(deadline)`

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#fe39c7fa-c5ac-4458-9ecd-8095632ed29b) |
<a name="fe39c7fa-c5ac-4458-9ecd-8095632ed29b"></a>
### 9324-728b18f/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31505
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b9d1a0f2cb6e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 21 seconds
(4/15) Running aws-sdk-ruby tests ... done in 15 seconds
(5/15) Running awscli tests ... done in 3 minutes and 33 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 6 minutes and 24 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 31 seconds
(9/15) Running minio-go tests ... done in 19 minutes and 8 seconds
(10/15) Running minio-java tests ... done in 26 minutes and 29 seconds
(11/15) Running minio-js tests ... FAILED in 3 minutes and 48 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-517d8697-ff6b-4b67-a42a-5b4ab0c0c3b9, recursive:true",
  "duration": 169,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/admin-handlers.go (hunk 4 above)
Comment (wlan0) [reply cmd/admin-handlers.go]: that would work too

Comment (wlan0) [reply cmd/admin-handlers.go]: Done

File: pkg/madmin/obd.go (hunk 3 above)
Comment (wlan0) [reply pkg/madmin/obd.go]: done

**Review by poornas — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#db77fb26-33b5-42ba-a5c4-ea2c36f5426b) |
<a name="db77fb26-33b5-42ba-a5c4-ea2c36f5426b"></a>
### 9324-20199f0/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30889
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8b7f3f39e8de:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 15 seconds
(5/15) Running awscli tests ... done in 3 minutes and 20 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 6 minutes and 19 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 32 seconds
(9/15) Running minio-go tests ... done in 20 minutes and 52 seconds
(10/15) Running minio-java tests ... done in 26 minutes and 59 seconds
(11/15) Running minio-js tests ... FAILED in 4 minutes and 25 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-e48a601b-6d97-4c02-9b27-8f1c672ba32a, recursive:true",
  "duration": 219,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Pull request merged**